File: structy/BinaryTree/leaf_list.py
class Node:
  def __init__(self, val):
    self.val = val
    self.left = None
    self.right = None

a = Node("a")
b = Node("b")
c = Node("c")
d = Node("d")
e = Node("e")
f = Node("f")

a.left = b
a.right = c
b.left = d
b.right = e
c.right = f

#      a
#    /   \
#   b     c
#  / \     \
# d   e     f

def leaf_list(root):
    if root is None:
        return []

    stack = [ root ]
    leaves = []

    while stack:
        current = stack.pop()

        if current.left is None and current.right is None:
            leaves.append(current.val)
        
        if current.right:
            stack.append(current.right)

        if current.left:
            stack.append(current.left)
    return leaves


print(leaf_list(a)) # -> [ 'd', 'e', 'f' ]

File: structy/BinaryTree/leafList.js
class Node {
  constructor(val) {
    this.val = val;
    this.left = null;
    this.right = null;
  }
}

const a = new Node("a");
const b = new Node("b");
const c = new Node("c");
const d = new Node("d");
const e = new Node("e");
const f = new Node("f");

a.left = b;
a.right = c;
b.left = d;
b.right = e;
c.right = f;

//      a
//    /   \
//   b     c
//  / \     \
// d   e     f


function leafList(root) {
    const leaves = []
    fillLeaves(root, leaves)
    return leaves
};

function fillLeaves(root, leaves) {
    if (!root) return;
    if (!root.right && !root.left) leaves.push(root.val)
    fillLeaves(root.left, leaves)
    fillLeaves(root.right, leaves)
};

console.log(leafList(a)); // -> [ 'd', 'e', 'f' ] 
